python/mxnet/optimizer.py

```diff
@@ -969,12 +969,21 @@ def sync_state_context(self, state, context):
 
     def set_states(self, states):
         """Sets updater states."""
-        self.states = pickle.loads(states)
+        states = pickle.loads(states)
+        if isinstance(states, tuple) and len(states) == 2:
```

Comment:
This is bad, add an argument

Reply:
It will become a bug very easily, very soon (if it is not already a bug) @piiswrong @szha

Reply:
Thanks for the comment. What's your suggestion @ptrendx?

Reply:
Not sure (especially since we need to keep backwards compatibility). Use dictionary instead of tuple for dumping both state and optimizer?

Reply:
That sounds more readable, though fundamentally the agreement on keys in dict isn't any stronger, more extensible, or more robust than type constraints and ordering of elements.

Reply:
That said, I'm open to a change with more thoughts to it.

Reply:
Optimizers already use tuples for states, so in that regard any type is better than tuple here. I would also argue that since fields in dictionary are named with human readable text it makes it much less error prone.

Reply:
Agreed. Proposed change in #7732